packages/store/ts/config/v2/table.ts

```diff
@@ -0,0 +1,145 @@
+  | SchemaInput<scope>
+  | keyof scope["validTypes"];
+
+export type TableFullInput<
+  schema extends SchemaInput<scope> = SchemaInput,
+  scope extends AbiTypeScope = AbiTypeScope,
+  primaryKey extends ValidKeys<schema, scope> = ValidKeys<schema, scope>
+> = {
+  schema: schema;
+  primaryKey: primaryKey;
+};
+
+// We don't use `conform` here because the restrictions we're imposing here are not native to typescript
+type validateTableShorthand<input, scope extends AbiTypeScope = AbiTypeScope> = input extends SchemaInput<scope>
+  ? // If a shorthand schema is provided, require it to have a static key field
+    "key" extends getStaticAbiTypeKeys<input, scope>
+    ? input
+    : NoStaticKeyFieldError
+  : input extends keyof scope["validTypes"]
+  ? input
+  : input extends string
+  ? keyof scope["validTypes"]
+  : SchemaInput<scope>;
+
+export type resolveTableShorthand<input, scope extends AbiTypeScope = AbiTypeScope> = input extends SchemaInput<scope>
+  ? "key" extends getStaticAbiTypeKeys<input, scope>
+    ? // If the shorthand includes a static field called `key`, use it as key
+      evaluate<TableFullInput<input, scope, ["key"]>>
+    : never
+  : input extends keyof scope["validTypes"]
+  ? evaluate<
+      TableFullInput<
+        { key: "bytes32"; value: input },
+        scope,
+        ["key" & getStaticAbiTypeKeys<{ key: "bytes32"; value: input }, scope>]
+      >
+    >
+  : never;
+
+export function resolveTableShorthand<input, scope extends AbiTypeScope = AbiTypeScope>(
+  input: validateTableShorthand<input, scope>,
+  scope?: scope
+): resolveTableShorthand<input, scope> {
+  // TODO: runtime implementation
+  return {} as never;
+}
+
+export type validateKeys<validKeys extends PropertyKey, keys> = {
+  [i in keyof keys]: keys[i] extends validKeys ? keys[i] : validKeys;
+};
+
+export function validateKeys<validKeys extends PropertyKey, keys = PropertyKey[]>(
+  keys: validateKeys<validKeys, keys>
+): keys {
+  // TODO: runtime implementation
+  return {} as never;
+}
+
+type validateTableFull<input, scope extends AbiTypeScope = AbiTypeScope> = input extends TableFullInput<
+  SchemaInput<scope>,
+  scope
+>
+  ? {
+      primaryKey: validateKeys<getStaticAbiTypeKeys<input["schema"], scope>, input["primaryKey"]>;
+      schema: input["schema"];
+    }
+  : input extends { primaryKey: unknown; schema: SchemaInput }
+  ? {
+      primaryKey: validateKeys<getStaticAbiTypeKeys<input["schema"], scope>, input["primaryKey"]>;
+      schema: SchemaInput<scope>;
+    }
+  : {
+      primaryKey: string[];
+      schema: SchemaInput<scope>;
+    };
+
+export type validateTableConfig<
+  input,
+  scope extends AbiTypeScope = AbiTypeScope
+> = input extends TableShorthandInput<scope>
+  ? validateTableShorthand<input, scope>
+  : input extends string
+  ? validateTableShorthand<input, scope>
+  : validateTableFull<input, scope>;
+
+type resolveTableFullConfig<
```

Comment:
should prob export ~everything?

Reply:
for this function i agree, in general i'm in favor of only exporting things that need to be exported so the non-exported things can be refactored more easily (bc we know nothing else depends on it)

Reply:
yeah makes sense, but I also know from working with viem that exporting intermediate type shapes is helpful when trying to extend/wrap things

but def wanna make sure we're ready to support whatever is exported

Reply:
exported most from this file